File: chapters/monte_carlo/code/R/monte_carlo.R

in_circle <- function(x, y){
        # Return True if the point is in the circle and False otherwise.
        radius = 1.0
        return((x*x + y*y) < radius*radius)
}

monte_carlo <- function(n_samples){
# Return the estimate of pi using the monte carlo algorithm.
        radius = 1.0
        
        # Sample x, y from the uniform distribution
        x <- runif(n_samples, 0, 1)
        y <- runif(n_samples, 0, 1)

        # Count the number of points inside the circle
        in_circle_count <- sum(in_circle(x, y))

        # Since we've generated points in upper left quadrant ([0,1], [0,1])
        # We need to multiply the number of points by 4	
        pi_estimate <- 4 * in_circle_count / (n_samples * radius * radius)

        return(pi_estimate)
}

pi_estimate <- monte_carlo(100000)
percent_error <- abs(pi - pi_estimate)/pi

print(paste("The estimate of pi is: ", formatC(pi_estimate)))
print(paste("The percent error is:: ", formatC(percent_error)))

File: chapters/monte_carlo/code/MonteCarlo.py
import math
import random


def in_circle(x, y, rad):
    if x**2 + y**2 < rad**2:
        return True
    else:
        return False


def monte_carlo_integration(sample, rad):
    count = 0
    for i in range(sample):
        x = random.random()*2 - 1
        y = random.random()*2 - 1
        if in_circle(x, y, rad):
            count += 1

    estimate = (4*count)/(sample*rad*rad)
    print("The Estimate of Pi is: %s" % estimate)
    print("The error is: %s" % (math.pi - estimate))


monte_carlo_integration(1000000, 1)
